src/sdk/python/tests/iotea/core/test_protocol_gateway.py

```diff
@@ -0,0 +1,251 @@
+        if asyncio.iscoroutinefunction(cb):
+            await cb(test_message, topic)
+        else:
+            cb(test_message, topic)
+
+    if group is None:
+        mocker.patch('src.iotea.core.util.mqtt_client.MqttProtocolAdapter.subscribe', wraps=mock_subscribe)
+        pg_subscribe_methods = [protocol_gateway.subscribe, protocol_gateway.subscribe_json]
+    else:
+        mocker.patch('src.iotea.core.util.mqtt_client.MqttProtocolAdapter.subscribe_shared',
+                     wraps=mock_subscribe_shared)
+        pg_subscribe_methods = [protocol_gateway.subscribe_shared, protocol_gateway.subscribe_json_shared]
+
+    callbacks = [async_callback, sync_callback]
+
+    for pg_subscribe in pg_subscribe_methods:
+        for callback in callbacks:
+            subscribe_options = ProtocolGateway.create_subscribe_options()
+            if group is None:
+                await pg_subscribe(test_topic, callback, subscribe_options=subscribe_options, force_wait=force_wait)
+                protocol_gateway.adapters[0].instance.subscribe.assert_called_with(test_topic, unittest.mock.ANY,
+                                                                                   subscribe_options)
+            else:
+                await pg_subscribe(group, test_topic, callback, subscribe_options=subscribe_options,
+                                   force_wait=force_wait)
+                protocol_gateway.adapters[0].instance.subscribe_shared.assert_called_with(group, test_topic,
+                                                                                          unittest.mock.ANY,
+                                                                                          subscribe_options)
+
+            # the subscription is not awaited but sent to event loop, so wait to get the job done
+            if not force_wait:
+                await asyncio.wait_for(callback_invoked_event.wait(), TIMEOUT)
+            # as Mock(async_callback) does not pass the test iscoroutingfunction,
+            # we need this hack to manually check if the callback has been invoked
+            test_case.assertEqual(True, callback_invoked_event.is_set(),
+                                  f'Subscribe callback {callback} was not invoked from {pg_subscribe}!')
+            callback_invoked_event.clear()
+
+
+# pylint: disable=redefined-outer-name
+def test_has_platform_adapter(test_case, protocol_gateway):
+    pg_config = ProtocolGateway.create_default_configuration([{'platform': True}])
+    test_case.assertEqual(True, protocol_gateway.has_platform_adapter(pg_config),
+                          'ProtocolGateway should have a platform adapter!')
+
+    pg_config = ProtocolGateway.create_default_configuration([{}])
+    test_case.assertEqual(False, protocol_gateway.has_platform_adapter(pg_config),
+                          'ProtocolGateway should not have a platform adapter!')
+
+
+@pytest.mark.parametrize('test_message', ['test message', {'json': 'test message'}])
```

Comment:
Having parametrizations and having two separate tests wrapped into one by a big if else block could be separated into two tests

Reply:
Refactored.